<?xml version="1.0" encoding="UTF-8"?>

<launch>

  <arg name="videopath"   default="" />
  <arg name="fps"         default="24" />
  <arg name="camera_name" default="camera_sim" />
  <arg name="camera_path" default="file://$(find visual_mtt2)/param/$(arg camera_name).yaml" />

  <arg name="bagpath"     default="" />
  <arg name="imgtopic"    default="/$(arg camera_name)/image_raw" />
  <arg name="compressed"  default="false" />
  <arg name="has_cinfo"   default="true" />
  <arg name="cimgtopic"   default="/image_converter/output_video" />

  <arg name="params"      default="default" />
  <arg name="tuning"      default="true" />
  <arg name="show_tracks" default="true" />

  <!-- videopath:     path to video file          (video only)              -->
  <!-- fps:           desired framerate           (video only)              -->
  <!-- camera_name:   camera name                 (video only)              -->
  <!-- camera_path:   path to camera .yaml        (video only)              -->

  <!-- bagpath:       path to recording           (rosbag only)             -->
  <!-- imgtopic:      topic from bag              (rosbag only)             -->
  <!-- compressed:    compressed topic from bag   (rosbag only)             -->
  <!-- has_cinfo:     camera info was recorded    (rosbag only)             -->
  <!-- cimgtopic:     compressed image topic      (compressed rosbag only)  -->

  <!-- params:        name of .yaml contained the desired parameter values  -->
  <!-- tuning:        enables dynamic reconfigure and displays measurements -->
  <!-- show_tracks:   enables display of tracking results                   -->

  <!-- #################################################################### -->

  <!-- play video and simulate camera -->
  <node unless="$(eval videopath == '')" pkg="visual_mtt2" type="video_player.py" name="player" output="screen">
    <param name="input" value="$(arg videopath)" />
    <param name="fps" value="$(arg fps)" />
    <remap from="output" to="/frame" />
  </node>
  <node unless="$(eval videopath == '')" pkg="visual_mtt2" type="camera_sim" name="$(arg camera_name)" output="screen">
    <param name="camera_name" value="$(arg camera_name)" />
    <param name="camera_info_path" value="$(arg camera_path)" />
    <remap from="input" to="/frame" />
  </node>

  <!-- #################################################################### -->

  <!-- play rosbag -->
  <param unless="$(eval bagpath == '')" name="use_sim_time" type="bool" value="true" />
  <node unless="$(eval bagpath == '')" pkg="rosbag" type="play" name="player" args="-l $(arg bagpath) --clock" launch-prefix="xterm -e" />

  <!-- decompress video if needed (rosbag only) -->
  <node if="$(arg compressed)"
    pkg="image_transport"
    type="republish"
    name="decompress"
    output="log"
    args="compressed in:=$(arg cimgtopic) out:=$(arg imgtopic)">
  </node>

  <!-- NEED TO SOLVE -->
  <!-- 1) problem when camera info isn't recorded -->
  <!-- 2) was recorded, but now on different namespace due to decompression -->

  <!-- #################################################################### -->

  <!-- launch visual_mtt -->
  <include file="$(find visual_mtt2)/launch/visual_mtt.launch">
    <arg name="params" value="$(arg params)" />
    <arg name="tuning" value="$(arg tuning)" />
    <arg name="show_tracks" value="$(arg show_tracks)" />
    <arg name="imgtopic" value="$(arg imgtopic)" />
  </include>


</launch>